addons/pos_loyalty/static/src/overrides/models/pos_order_line.js

```diff
@@ -33,32 +63,48 @@ patch(PosOrderline.prototype, {
         }
         this.giftBarcode = json.giftBarcode;
         this.giftCardId = json.giftCardId;
-        this.eWalletGiftCardProgram = this.models["loyalty.program"].get(
-            json.eWalletGiftCardProgramId
-        );
         super.init_from_JSON(...arguments);
     },
+    serialize() {
+        const isNegativeCoupon = this.coupon_id?.id < 0;
+        const json = super.serialize(...arguments);
+        if (isNegativeCoupon) {
+            json.coupon_id = undefined;
+        }
+        return json;
+    },
+    setOptions(options) {
+        if (options.eWalletGiftCardProgram) {
+            this.update({ e_wallet_program_id: options.eWalletGiftCardProgram });
+        }
+        if (options.giftBarcode) {
+            this.update({ gift_barcode: options.giftBarcode });
+        }
+        if (options.giftCardId) {
+            this.update({ gift_card_id: this.models["loyalty.card"].get(options.giftCardId) });
+        }
+        return super.setOptions(...arguments);
+    },
     getEWalletGiftCardProgramType() {
-        return this.eWalletGiftCardProgram && this.eWalletGiftCardProgram.program_type;
+        return this.e_wallet_program_id && this.e_wallet_program_id.program_type;
     },
     ignoreLoyaltyPoints({ program }) {
         return (
             ["gift_card", "ewallet"].includes(program.program_type) &&
-            this.eWalletGiftCardProgram &&
-            this.eWalletGiftCardProgram.id !== program.id
+            this.e_wallet_program_id?.id !== program.id
         );
     },
     isGiftCardOrEWalletReward() {
-        const coupon = this.pos.couponCache[this.coupon_id];
+        const coupon = this.coupon_id;
         if (!coupon || !this.is_reward_line) {
             return false;
         }
-        const program = this.models["loyalty.program"].get(coupon.program_id);
-        return ["ewallet", "gift_card"].includes(program.program_type);
+        return ["ewallet", "gift_card"].includes(coupon.program_id?.program_type);
     },
     getGiftCardOrEWalletBalance() {
-        const coupon = this.pos.couponCache[this.coupon_id];
-        return this.env.utils.formatCurrency(coupon?.balance || 0);
+        const coupon = this.coupon_id;
+        // TODO: find a better way to access `env.utils` from the model instances.
+        return this.pos.env.utils.formatCurrency(coupon?.points || 0);
```

Comment:
There is a formatCurrency methods in `odoo/addons/point_of_sale/static/src/app/models/utils/currency.js`

This requires more arguments, but allows you to access it without pos_store or other dependencies.